### ISSUE TITLE
query: potential candidate for query format: jsonpath

### ISSUE THREAD

**Issue by reschke** [httpwg/http-extensions#2924]
Spec: https://www.rfc-editor.org/rfc/rfc9535.html

- how to augment the existing media type definition for QUERY (do we need to?)
- what response format to use? (application/json)?

**Comment by timbray** [comment]
If I'm building an API around the new QUERY method, surely I'd have to specify the media types for both query and response, right? So I'm not sure what's special about JSONPath.

**Comment by reschke** [comment]
The theory is that for a media type to be applicable for QUERY the semantics for QUERY should be specified somewhere. That idea dates back to the time when people just used application/json as payload format for PATCH, with different ideas what that actually means.

But I agree that the semantics for JsonPath with QUERY are pretty clear, maybe except for the question what the JSON document being queried actually is. But that can remain an implementation detail of the server. Any recommendations about what format we should use in the response? Maybe you even could suggest an example?

**Comment by timbray** [comment]
I disagree at a basic level. A JSONPath selects the parts of  some JSON
text that it matches.  What the JSON text is, and what the effect of
selecting parts of it is, seem entirely application-dependent.  Maybe it’s
applied to the user’s order history and forwards the selected portions to
Accounting, and returns status code 200 but no body.  Maybe it applies the
JSONPath to that same order history, creates an analysis, and returns
application/csv.  These things are designed to be general-purpose
expressions and I really can’t think of anything useful that could be said.

 -T

**Comment by reschke** [comment]
Ok, just checking: we don't want to define a standard response format, because we want to leave it specific to the server? I agree with that.

But for an example, would you recommend a specific response format (iff that this is just an example)?

**Comment by timbray** [comment]
That example of sending a JSONPath which is applied to the user's saved
data and returns an analysis in a  CSV isn't terrible? Or it might return a
nice graph as a JPEG or PDF.

**Comment by mnot** [comment]
I think the difference here is that PATCH has side effects on the resource, whereas QUERY should not (generally). PATCH needed definition to describe those side effects, so it doesn't become another POST escape hatch.

**Comment by reschke** [comment]
Ok, so we'll just add an example using jsonpath.

**Comment by reschke** [comment]
Proposal:

~~~
QUERY /errata.json HTTP/1.1
Host: example.org
Content-Type: example/query
Accept: application/jsonpath

$..[?@doc-id=="RFC9110" && @errata_status_code!="Rejected"].[errata_id, submit_date]
~~~

from a fictious RFC errata database, selects all errata for RFC9110 which have not been rejected, and returns id, submission_date for each

~~~
HTTP/1.1 200 OK
Content-Type: application/json

[
  "7105",
  "2022-08-26",
  "7109",
  "2022-08-31",
  "7138",
  "2022-09-23",
  "7306",
  "2023-01-13",
  "7419",
  "2023-04-11"
]
~~~

**Comment by timbray** [comment]
On Nov 1, 2024 at 1:34:45 PM, Julian Reschke ***@***.***>
wrote:

Content-Type: example/query
>
Accept: application/jsonpath
>

That should be Content-Type: application/jsonpath, Accept:
application/json, no?

Proposal:

**Comment by reschke** [comment]
Oh, indeed. Thanks for catching this. Will edit in place.

**Comment by reschke** [comment]
We could also use this as example of conneg:

~~~
HTTP/1.1 200 OK
Content-Type: text/csv

errata_id, submit_date
7105, 2022-08-26,
7109, 2022-08-31,
7138, 2022-09-23,
7306, 2023-01-13,
7419, 2023-04-11
~~~

**Comment by mnot** [comment]
Part of #1906

**Comment by cabo** [comment]
I agree with using Accept: to specify the desired response format (we could even give application/json a specific meaning in this context); application/jsonpath is fine as a query content type.

**Comment by reschke** [comment]
@timbray @cabo - I need help with getting the syntax correct. What tool should I use? Is https://jsonpath.com/ ok? FWIW, the sample data is https://www.rfc-editor.org/errata.json

**Comment by reschke** [comment]
-> #1906

**Comment by cabo** [comment]
> [@timbray](https://github.com/timbray) [@cabo](https://github.com/cabo) - I need help with getting the syntax correct. What tool should I use? Is https://jsonpath.com/ ok? FWIW, the sample data is https://www.rfc-editor.org/errata.json

I use a simple tool called jpt (gem install jpt).  @glyn probably has more recommendations...
(See also his blog article https://underlap.org/jsonpath-rfc-nearing-publication .)

**Comment by reschke** [comment]
jpt seems to use the same parser as https://jsonpath.com/...

I should have been more precise on what problems I encountered:

Escaping "-" in names. The query

`$..[?@.doc-id=='RFC4954']`

results in parse error:

~~~
Expected one of [A-Z], [a-z], "_", [-퟿], [-􏿿], [0-9], " ", "\t", "\n", "\r", "[", ".", "==", "!=", "<=", ">=", "<", ">", "..", "&&", "||", ",", "]" at line 1, column 11 (byte 11):
~~~

The other question was how to specify multiple members for output; in the original example I had:

`$..[?@doc-id=="RFC9110" && @errata_status_code!="Rejected"].[errata_id, submit_date]`

Is this even possible in regular JSONPath?

**Comment by martinthomson** [comment]
Here's what I found works if I altered their example: `$["doc-id"]` or `$["doc\u002did"]`

That is, if you want "-", you have to not use the `.<label>` syntax, but instead `["<label>"]`, using JSON-style quoting in that string.

**Comment by cabo** [comment]
> jpt seems to use the same parser as https://jsonpath.com/...
> 
> I should have been more precise on what problems I encountered:
> 
> Escaping "-" in names. The query
> 
> `$..[?@.doc-id=='RFC4954']`

@["doc-id"] is a way to include other characters in a name.

> The other question was how to specify multiple members for output; in the original example I had:
> 
> `$..[?@doc-id=="RFC9110" && @errata_status_code!="Rejected"].[errata_id, submit_date]`
> 
> Is this even possible in regular JSONPath?

We didn't include projection (the SELECT... part of SQL) in JSONPath.
You get pointers into an instance, or the values at these points.
No construction of new values.

**Comment by reschke** [comment]
I now have https://github.com/httpwg/http-extensions/blob/7383116f9ebbfabbea128f8e0519412dc966e27b/draft-ietf-httpbis-safe-method-w-body.xml#L868

Do you have a good idea how to produce JSONPath results for the response body? Is there a notation I could use?

**Comment by cabo** [comment]
JSONPath does not specify an output format, just saying that this could be a list of normalized paths and/or the values at these paths.  The reply at line 896 looks good to me as an example for representing the latter list as a JSON array.

Please do fix line 884 ["doc\u002did"]; I have no idea what made Martin escape the hyphen-minus, which is perfectly valid ASCII at this point.
Also, the line breaking and indentation of the query is quite weird.

**Comment by reschke** [comment]
Please try. "doc-id" is not accepted by the parser: https://github.com/httpwg/http-extensions/issues/2924#issuecomment-2670788751

For the query, I wanted to indent it in order to avoid overflowing the rfc txt limits. If you can suggest a better variant, please go ahead.

**Comment by cabo** [comment]
```
   name-segment        = ("[" name-selector "]") /
                         ("." member-name-shorthand)
   member-name-shorthand = name-first *name-char
   name-selector       = string-literal
   string-literal      = %x22 *double-quoted %x22 /     ; "string"
                         %x27 *single-quoted %x27       ; 'string'
```
hyphen-minus is not a name-char, so a member-name-shorthand like

```
@.doc-id
```

is a syntax error, as is

```
@.doc\u002did
```

Instead,

```
@["doc-id"]
```

or

```
@['doc-id']
```

is the correct way to use this member name in a name-segment.
The string-literals inside a name-selector also allow escapes like \u002d, but there is no point for hyphen-minus.

**Comment by reschke** [comment]
That's why I was so confused. Let me try again.

**Comment by glyn** [comment]
> > [@timbray](https://github.com/timbray) [@cabo](https://github.com/cabo) - I need help with getting the syntax correct. What tool should I use? Is https://jsonpath.com/ ok? FWIW, the sample data is https://www.rfc-editor.org/errata.json
> 
> I use a simple tool called jpt (gem install jpt). [@glyn](https://github.com/glyn) probably has more recommendations... (See also his blog article https://underlap.org/jsonpath-rfc-nearing-publication .)

Sorry for the slow reply. Here are two "live" implementations of RFC 9535 which you may already have gotten from my blog post:
* https://json-everything.net/json-path
* https://serdejsonpath.live/

**Comment by reschke** [comment]
See https://github.com/httpwg/http-extensions/pull/3002

**Comment by reschke** [comment]
@glyn - thanks; I just merged the new examples section. Maybe you want to check?

**Comment by glyn** [comment]
@reschke:

> [@glyn](https://github.com/glyn) - thanks; I just merged the new examples section. Maybe you want to check?

I skimmed through https://github.com/httpwg/http-extensions/pull/3002, but wasn't really sure what which JSONPath query expressions you wanted me to check. If you can drop them here, I'll take a look.

**Comment by reschke** [comment]
---> https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-08.html#name-more-query-formats

**Comment by glyn** [comment]
~~~
$..[
     ?@.errata_status_code=="Rejected"
     && @.submit_date>"2024"
   ]
   ["doc-id"]
~~~
Confirmed that this is valid JSONPath syntax.
